src/Compressors/sparse_sign.jl

```
@@ -0,0 +1,238 @@
+    n_rows::Int64
+    n_cols::Int64
+    nnz::Int64
+    scale::Vector{Number}
+    op::SparseMatrixCSC
+end
+
+function complete_compressor(ingredients::SparseSign, A::AbstractMatrix)
+    if ingredients.cardinality == Left
+        n_rows = ingredients.compression_dim
+        n_cols = size(A, 1)
+        initial_size = n_cols
+    else
+        n_rows = size(A, 2)
+        n_cols = ingredients.compression_dim
+        initial_size = n_rows
+    end
+    nnz = ingredients.nnz
+    compression_dim = ingredients.compression_dim
+    # get the element type of the matrix
+    T = eltype(A)
+    total_nnz = initial_size * nnz
+    idxs = Vector{Int64}(undef, total_nnz)
+    start = 1
+    for i in 1:initial_size
+        # every grouping of nnz entries corresponds to each row/column in sample
+        stop = start + nnz - 1
+        # Sample indices from the intial_size
+        idx_view = view(idxs, start:stop)
+        sample!(
+            1:compression_dim, 
+            idx_view, 
+            replace = false, 
+            ordered = true
+        )
+        start = stop + 1
+    end
+    
+    # store the number in the type equivalent to the matrix A
+    sc = convert(T, 1 / sqrt(nnz))
+    # store as a vector to prevent reallocation during upadate
+    scale = [-sc, sc]
+    signs = rand(scale, total_nnz)  
+    ptr = collect(1:nnz:total_nnz+1)  
+
+    #If left sketching store as a sparse matrix if right store as sparse matrix transpose
+    if ingredients.cardinality != Left
+        op = SparseMatrixCSC{T, Int64}(n_cols, n_rows, ptr, idxs, signs)'
+    else
+        op = SparseMatrixCSC{T, Int64}(n_rows, n_cols, ptr, idxs, signs)
+    end
+
+    return SparseSignRecipe(ingredients.cardinality, n_rows, n_cols, nnz, scale, op)
+end
+
+# This runs sparse Sign with both a matrix and vector input (for linear solver)
+function complete_compressor(ingredients::SparseSign, A::AbstractMatrix, b::AbstractVector)
+    complete_compressor(ingredients, A)
+end
```

I do not think you need this function here as `complete_compressor(ingredients, A, b)` should be handled by the default one in `Compressors.jl`

---

This might be a good developer note: only the necessary `complete_compressor` needs to be defined. So if (ingredients, A, b) is defined then it would imply that (ingredients, A) would be insufficient to complete a compressor and should throw the error (which is handled by the function in Compressors.jl).

---

This is a good point.